File: 2D-dispersion_alex/get_duddesign.m
function [duddesign,domegaddesign] = get_duddesign(Kr,Mr,omega,u,dKrddesign,dMrddesign)
    matrix_size = 2*size(Kr,1) + 2;
    A_sub1 = Kr-omega^2*Mr;
    A_sub2 = -Mr*u;
    A_sub3 = u'*Mr;
    z1 = zeros(1,size(u',2));
    z2 = zeros(1,size(u',2)-1);
    if matrix_size < 2000
%         A = full([real(Kr-omega^2*Mr) -imag(Kr-omega^2*Mr) real(-Mr*u) -imag(-Mr*u);
%             imag(Kr-omega^2*Mr)  real(Kr-omega^2*Mr) imag(-Mr*u)  real(-Mr*u);
%             real(u'*Mr)         -imag(u'*Mr)         0            0          ;
%             zeros(1,size(u',2))  1e3 zeros(1,size(u',2)-1) 0       0        ]);
        
        A = full([real(A_sub1) -imag(A_sub1) real(A_sub2) -imag(A_sub2);
                  imag(A_sub1)  real(A_sub1) imag(A_sub2)  real(A_sub2);
                  real(A_sub3) -imag(A_sub3)      0             0      ;
                  z1            1e3 z2            0             0      ]);
        
%         b = full([
%             real(-(dKrddesign-omega^2*dMrddesign)*u);
%             imag(-(dKrddesign-omega^2*dMrddesign)*u);
%             real(-1/2*u'*dMrddesign*u);
%             0
%             ]);
    else
        A = [real(Kr-omega^2*Mr) -imag(Kr-omega^2*Mr) real(-Mr*u) -imag(-Mr*u);
            imag(Kr-omega^2*Mr)  real(Kr-omega^2*Mr) imag(-Mr*u)  real(-Mr*u);
            real(u'*Mr)         -imag(u'*Mr)         0            0          ;
            zeros(1,size(u',2))  1e3 zeros(1,size(u',2)-1) 0       0        ];
        
%         b = [
%             real(-(dKrddesign-omega^2*dMrddesign)*u);
%             imag(-(dKrddesign-omega^2*dMrddesign)*u);
%             real(-1/2*u'*dMrddesign*u);
%             0
%             ];
    end
    
%         b = [
%             real(-(dKrddesign-omega^2*dMrddesign)*u);
%             imag(-(dKrddesign-omega^2*dMrddesign)*u);
%             real(-1/2*u'*dMrddesign*u);
%             0
%             ];
        
    b_sub1 = -(dKrddesign-omega^2*dMrddesign)*u;
    b_sub2 = -1/2*u'*dMrddesign*u;

    b = [
        real(b_sub1);
        imag(b_sub1);
        real(b_sub2);
        0
        ];
    
    x = A\b;
    %     if rank(A)<size(A,1) %|| rcond(A)<1e-8
    %         warning('duddesign system is rank deficient')
    %     end
    duddesign = x(1:size(u,1)) + 1i*x(size(u,1)+1:end-2);
    domegaddesign = x(end-1) + 1i*x(end);
end

File: 2d-dispersion-py/get_duddesign.py
"""
Eigenvector derivative calculation.

This module contains functions for computing derivatives of eigenvectors
with respect to design parameters.
"""

import numpy as np
from scipy.linalg import solve


def get_duddesign(Kr, Mr, omega, u, dKrddesign, dMrddesign):
    """
    Compute derivative of eigenvector with respect to design parameter.
    
    Parameters
    ----------
    Kr : array_like
        Reduced stiffness matrix
    Mr : array_like
        Reduced mass matrix
    omega : float
        Angular frequency
    u : array_like
        Eigenvector
    dKrddesign : array_like
        Derivative of reduced stiffness matrix with respect to design parameter
    dMrddesign : array_like
        Derivative of reduced mass matrix with respect to design parameter
        
    Returns
    -------
    duddesign : array_like
        Derivative of eigenvector with respect to design parameter
    domegaddesign : complex
        Derivative of frequency with respect to design parameter
    """
    
    matrix_size = 2 * Kr.shape[0] + 2
    
    # Build the augmented system matrix
    A_sub1 = Kr - omega**2 * Mr
    A_sub2 = -Mr @ u
    A_sub3 = u.conj().T @ Mr
    z1 = np.zeros((1, u.shape[0]))
    z2 = np.zeros((1, u.shape[0] - 1))
    
    if matrix_size < 2000:
        # Use dense matrices for small systems
        A = np.block([
            [np.real(A_sub1), -np.imag(A_sub1), np.real(A_sub2), -np.imag(A_sub2)],
            [np.imag(A_sub1),  np.real(A_sub1), np.imag(A_sub2),  np.real(A_sub2)],
            [np.real(A_sub3), -np.imag(A_sub3),      0,             0],
            [z1,              1e3,            z2,             0]
        ])
    else:
        # Use sparse matrices for large systems
        from scipy.sparse import bmat
        A = bmat([
            [np.real(A_sub1), -np.imag(A_sub1), np.real(A_sub2), -np.imag(A_sub2)],
            [np.imag(A_sub1),  np.real(A_sub1), np.imag(A_sub2),  np.real(A_sub2)],
            [np.real(A_sub3), -np.imag(A_sub3),      0,             0],
            [z1,              1e3,            z2,             0]
        ])
    
    # Build right-hand side
    b_sub1 = -(dKrddesign - omega**2 * dMrddesign) @ u
    b_sub2 = -1/2 * u.conj().T @ dMrddesign @ u
    
    b = np.concatenate([
        np.real(b_sub1),
        np.imag(b_sub1),
        np.real(b_sub2),
        [0]
    ])
    
    # Solve the system
    x = solve(A, b)
    
    # Extract eigenvector and frequency derivatives
    n = u.shape[0]
    duddesign = x[:n] + 1j * x[n:2*n]
    domegaddesign = x[-2] + 1j * x[-1]
    
    return duddesign, domegaddesign

